File: plugins/capawesome-capacitor-printer/ios/Plugin/PrinterPlugin.swift
import Foundation
import Capacitor

@objc(PrinterPlugin)
public class PrinterPlugin: CAPPlugin {
    public let tag = "FirebaseRemoteConfig"
    public let errorHtmlMissing = "html must be provided."
    public let errorPathMissing = "path must be provided."
    public let errorFileNotExist = "File does not exist."

    private var implementation: Printer?

    override public func load() {
        self.implementation = Printer(plugin: self)
    }

    @objc func printPdf(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? "Document"
        guard let path = call.getString("path") else {
            call.reject(errorPathMissing)
            return
        }

        guard let url = implementation?.getFileUrlByPath(path) else {
            call.reject(errorFileNotExist)
            return
        }

        let options = PrintPdfOptions(name: name, url: url)

        implementation?.printPdf(options, completion: { error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription)
                return
            }
            call.resolve()
        })
    }

    @objc func printHtml(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? "Document"
        guard let html = call.getString("html") else {
            call.reject(errorHtmlMissing)
            return
        }

        let options = PrintHtmlOptions(name: name, html: html)

        implementation?.printHtml(options, completion: { error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription)
                return
            }
            call.resolve()
        })
    }

    @objc func printWebView(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? "Document"

        let options = PrintWebViewOptions(name: name)

        implementation?.printWebView(options, completion: { error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription)
                return
            }
            call.resolve()
        })
    }
}

File: plugins/capawesome-capacitor-printer/android/src/main/java/io/capawesome/capacitorjs/plugins/printer/PrinterPlugin.java
package io.capawesome.capacitorjs.plugins.printer;

import android.net.Uri;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.printer.classes.options.PrintHtmlOptions;
import io.capawesome.capacitorjs.plugins.printer.classes.options.PrintPdfOptions;
import io.capawesome.capacitorjs.plugins.printer.classes.options.PrintWebViewOptions;
import io.capawesome.capacitorjs.plugins.printer.interfaces.EmptyResultCallback;

@CapacitorPlugin(name = "Printer")
public class PrinterPlugin extends Plugin {

    public static final String TAG = "PrinterPlugin";
    public static final String ERROR_HTML_MISSING = "html must be provided.";
    public static final String ERROR_PATH_MISSING = "path must be provided.";
    public static final String ERROR_FILE_NOT_EXIST = "File does not exist.";

    private Printer implementation;

    @Override
    public void load() {
        implementation = new Printer(this);
    }

    @PluginMethod
    public void printPdf(PluginCall call) {
        try {
            String name = call.getString("name", "Document");
            String path = call.getString("path");
            if (path == null) {
                call.reject(ERROR_PATH_MISSING);
                return;
            }

            Uri uri = implementation.getUriByPath(path);
            boolean fileExists = implementation.isFileExists(uri);
            if (!fileExists) {
                call.reject(ERROR_FILE_NOT_EXIST);
                return;
            }

            PrintPdfOptions options = new PrintPdfOptions(name, uri);
            EmptyResultCallback callback = new EmptyResultCallback() {
                @Override
                public void success() {
                    call.resolve();
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.printPdf(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void printHtml(PluginCall call) {
        try {
            String name = call.getString("name", "Document");
            String html = call.getString("html");
            if (html == null) {
                call.reject(ERROR_HTML_MISSING);
                return;
            }

            PrintHtmlOptions options = new PrintHtmlOptions(name, html);
            EmptyResultCallback callback = new EmptyResultCallback() {
                @Override
                public void success() {
                    call.resolve();
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.printHtml(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void printWebView(PluginCall call) {
        try {
            String name = call.getString("name", "Document");

            PrintWebViewOptions options = new PrintWebViewOptions(name);
            EmptyResultCallback callback = new EmptyResultCallback() {
                @Override
                public void success() {
                    call.resolve();
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.printWebView(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }
}
